scripts/mockgen.sh

```diff
@@ -1,8 +1,19 @@
 #!/usr/bin/env bash
 
-# Use this script to regererate generated mock files
+# Use this script to regenerate generated mock files
 # after changing signatures of interfaces in these packages
 
+# check if mockgen is installed
+mockgen --version
+if [ $? -eq 0 ]; then
+  echo "✓ mockgen is installed"
+else
+  echo "mockgen not found; installing it..."
+  go install github.com/golang/mock/mockgen@v1.6.0
```

Comment:
Could we include this in the `goget-tools` `Make` target? I find it a bit confusing that the script also installs `mockgen`.. At least, we could display a message instructing to install `mockgen` with `make goget-tools` if it is not installed or if it does not have the right version.
Also, I see that we now have 2 versions that will need to be kept in sync (the library in `go.mod` and the version of the `mockgen` binary).
Would it be possible to have a single source (for future maintenance)?

Reply:
With hindsight, I think it would be more maintainable if we could make the script detect the appropriate version from `go.mod` using `go mod graph` and use it with `go run -mod=readonly` for example; this way, we won't even need to `go install` it, and our source of truth is `go.mod` (well, how the dependency graph is actually resolved by `go`).

Quick example of how we could do this:

```bash
set -x

gomockCoordinates=$(go mod graph | grep '^github.com/redhat-developer/odo\sgithub.com/golang/mock@' | awk '{print $2}')

mockgen="go run -mod=readonly $(echo "$gomockCoordinates" | sed 's/\/mock@/\/mock\/mockgen@/g')"

# Now we can use $mockgen as our command, e.g.:
$mockgen -source=pkg/kclient/interface.go \
    -package kclient \
    -destination pkg/kclient/mock_Client.go

...
```

@feloy @valaparthvi What do you think?

Reply:
This solution introduces many dependencies (every time I see greps and seds I know something will break if a space changes somewhere :( )

I would prefer to duplicate the version in go.mod and Makefile (by installing the binary) instead

Reply:
> This solution introduces many dependencies (every time I see greps and seds I know something will break if a space changes somewhere :( )
> 
:) Okay, I agree using `grep` and `sed` is a bit fragile, but I wanted to discuss the overall idea.

How about that one using `go list`. I stumbled upon this, to extract the version in a much simpler way:
```
$ go list -mod=readonly -m -f '{{.Version}}' github.com/golang/mock
v1.6.0
```

So we could simply use:
```
gomockVersion=$(go list -mod=readonly -m -f '{{.Version}}' github.com/golang/mock)
mockgen="go run -mod=readonly github.com/golang/mock/mockgen@$gomockVersion"

echo Using mockgen "$($mockgen --version)"
```

> I would prefer to duplicate the version in go.mod and Makefile (by installing the binary) instead
>
If we go that way, I'd suggest documenting both in the Makefile and `go.mod` that we need to update both places.
Even there, I think we might still be out of sync in case another dependency requires a higher version. For example, if I downgrade `github.com/golang/mock` to `v1.5.0` in `go.mod`, `go mod` will automatically set it to `v1.6.0` because one of our dependencies requires a higher version (`v1.6.0`).

Reply:
I'm not sure that the important thing is that we use the same version between binary and library.

The problem we have currently is that different developers have different version of the binary, and the mocks files are modified endlessly because of version change depending on who re-runs the script.

Reply:
> I'm not sure the important thing is to use the same version between binary and library.
> 
> The problem we have currently is that different developers have different version of the binary, and the mocks files are modified endlessly because of version change depending on who re-runs the script.
>

In [their Readme](https://github.com/golang/mock#readme), they state that "You should try to keep the library in sync with the version of mockgen used to generate your mocks."; that's why I thought ensuring everyone is using the right version while not needing to install mockgen would probably make sense.

But okay, no problem, it's fine if we duplicate the version in both `go.mod` and `Makefile`.

Reply:
> ```
> gomockVersion=$(go list -mod=readonly -m -f '{{.Version}}' github.com/golang/mock)
> mockgen="go run -mod=readonly github.com/golang/mock/mockgen@$gomockVersion"
> 
> echo Using mockgen "$($mockgen --version)"
> ```

I liked the solution Armel suggested in https://github.com/redhat-developer/odo/pull/6604#discussion_r1114053184, so I have modified the mockgen script to use the library version and `go run -mod=readonly` to run the scripts.

However, I have also added instructions to install mockgen script as a part of the goget-tools target in case someone wants to run mockgen on a single file. Here as well, we make sure we fetch the library version and use it.